### ISSUE TITLE
1.12.2 update to add 2-dimensional grid support for machine generation

### ISSUE THREAD

**Issue by SizableShrimp** [CompactMods/CompactMachines#429]
I have tried contacting the mod creator, @thraaawn aka Davenonymous, on Discord and Twitter and have no received any word. I would like to release a new update to the mod that is mainly a huge Quality of Life feature update.

This update will mainly focus on turning the mod into a grid-based system instead of only operating on the x-axis. I have already refactored most of the codebase to work with this and am mainly in a bug-testing state. I also have made sure to make it **completely backwards-compatible** and am actively testing this compatability as well. So far it has worked with pretty good success. I have also refactored the config file and config GUI to have better localization and more comments on what each values do, along with making it look nicer in general. This update is in the works, but I am confident it will be a nice feature because a big drawback to the mod is the rendering issues that happen at millions of blocks out. On a large server, this is easy to get machines this far out and it is really detrimental to the player experience. I would also be happy to take over the project if thraaawn does not have the time to be able to update it. I hope this reaches the mod creator so that this update can get out to people. I would really enjoy having a conversation with the mod author. Currently, I am not experience much in Forge mod development but have many years of experience in plugin development and Java. Therefore, I am working mainly on server-based features that do not involve rendering so this update will not feature any new blocks/items. Please feel free to comment below.

**Comment by 4rz0** [comment]
Have you tried submitting a PR for those changes?
I can't find any with your name on it.

**Comment by SizableShrimp** [comment]
Sorry, I never actually pushed it to my fork. It is for 1.12.2, and could be the final update before the 1.16.4 update is released. Not sure if you want to take it at all at this point. I will push it to my fork now and make a PR. Since much of the 1.12 version of the mod is based on using `mod 1024` to do rendering and find grid locations, there could still be bugs that I have missed. I have tried to modify as little as possible to make it compatible so as to break less stuff. Most of the bugs I have found (and fixed) so far have been related to NPEs because of the code not having access to the id -> grid location map in the context of the code, or this map not being created yet. I will look through the code and manually try to find NPEs before making a PR, and I will add a lot more info to that PR.
